README.md
# balena audio block

Provides an easy way to work with audio applications in a containerized environment.
The `audio` block is a docker image that runs a [PulseAudio](https://www.freedesktop.org/wiki/Software/PulseAudio/) server optimized for balenaOS.

## Features

- PulseAudio v15 audio server
- Configuration optimized for balenaOS, extendable via PA config files
- Supports both TCP and UNIX socket communication
- Bluetooth and ALSA support out of the box
- Companion library to send PA commands and handle events using JavaScript

## Usage

#### Prebuilt images

We maintain images for this block on balenaHub Container Registry. The images can be accessed using:

`bh.cr/balenalabs/audio-<arch>` or `bhcr.io/balenalabs/audio-<arch>` where `<arch>` is one of: `rpi`, `armv7hf`, `aarch64` or `amd64`.

For details on how to select a specific version or commit version of the image see our [documentation](https://github.com/balena-io/open-balena-registry-proxy/#usage).

#### docker-compose file
To use this image, create a container in your `docker-compose.yml` file as shown below:

```yaml
version: '2'

volumes:
  pulse:                          # Only required if using PA over UNIX socket

services:

  audio:
    image: bh.cr/balenalabs/audio-<arch> # where <arch> is one of rpi, armv7hf, aarch64 or amd64
    privileged: true
    labels:
      io.balena.features.dbus: 1  # Only required for bluetooth support
    ports:
      - 4317:4317                 # Only required if using PA over TCP socket
    volumes:
      - 'pulse:/run/pulse'        # Only required if using PA over UNIX socket

  my-audio-app:
    build: ./my-audio-app
    volumes:
      - 'pulse:/run/pulse'        # Only required if using PA over UNIX socket
```

#### Send/receive audio

In order to route audio through the `audio` block there are a few environment variables you'll need to set. Note that they must be set on your client container, where your audio application is running and **not** on the block itself. We recommend setting them in the `Dockerfile`. 

| Environment variable | Description |
| --- | --- |
| `PULSE_SERVER` | **Required** Address of the PulseAudio server which you want to connect to. Depending on the communication protocol you want to use it can be: <br>- *UNIX socket*: `PULSE_SERVER=unix:/run/pulse/pulseaudio.socket`<br>- *TCP socket*: `PULSE_SERVER=tcp:audio:4317` |
| `PULSE_SINK` | **Optional** The PulseAudio sink your application will send audio to. If not set, the block will use the PulseAudio default sink. Unless you are building a complex audio application we don't recommend setting this variable. If you want to select which output to use, for example HDMI or audio jack for a Raspberry Pi use the `AUDIO_OUTPUT` [env var](#environment-variables) on the block to select the output device.  |
| `PULSE_SOURCE` | **Optional** The PulseAudio source your application will get audio from. |

Setting these environment variables will instruct your application to route audio to the PulseAudio server on the `audio` container. For this to work your application must have built-in support for  PulseAudio as an audio backend. Most applications do, though some might require installing or configuring additional packages. If your application does not have native support for the PulseAudio backend you'll need to use your container's ALSA backend to bridge over to PulseAudio.

Read on for details on both alternatives. We've also included some examples in the `examples` folder (along with the `docker-compose.yml` file) so be sure to check that as well for implementation details.

**PulseAudio backend**

For applications with PulseAudio support, the audio is routed as follows: 

`[client-container] audio-app --> [audio] PulseAudio --> [audio] ALSA --> Audio Hardware`

Here is a non-exhaustive list of applications with PulseAudio backend that have been tested to work, feel free to PR more: 
- [SoX](http://sox.sourceforge.net/): PA backend distributed via `libsox-fmt-pulse` package
- [MPlayer](http://www.mplayerhq.hu/): Native PA backend
- [FFmpeg](https://ffmpeg.org/): Native PA backend

**ALSA bridge**

For audio applications that don't have built-in PulseAudio support you can use ALSA to brige the gap:

`[client-container] audio-app --> [client-container] ALSA --> [audio] PulseAudio --> [audio] ALSA --> Audio Hardware`

Setting up the ALSA bridge requires extra configuration steps on your containers so we created a few bash scripts to simplify the process:

- [Debian](scripts/alsa-bridge/debian-setup.sh)
- [Alpine](scripts/alsa-bridge/alpine-setup.sh) 

Before making use of audio capabilities you should run this script. An easy way to do so is by including the following instruction in your `Dockerfile`:

```Dockerfile
RUN curl -skL https://raw.githubusercontent.com/balena-labs-projects/audio/master/scripts/alsa-bridge/debian-setup.sh| sh
```

## Customization
### Extend image configuration

You can extend the `audio` block to include custom configuration as you would with any other `Dockerfile`. Just make sure you don't override the `ENTRYPOINT` as it contains important system configuration.

Here are some of the most common extension cases: 

- Pass a flag to the PulseAudio server:

```Dockerfile
FROM bh.cr/balenalabs/audio-aarch64

CMD [ "--disallow-module-loading" ]
```

- Add custom configuration files:

```Dockerfile
FROM bh.cr/balenalabs/audio-aarch64

COPY custom.pa /usr/src/custom.pa
CMD [ "pulseaudio", "--file /usr/src/custom.pa" ]
```

- Start PulseAudio from your own bash script:

```Dockerfile
FROM bh.cr/balenalabs/audio-aarch64

COPY custom.pa /usr/src/custom.pa
COPY start.sh /usr/src/start.sh
CMD [ "/bin/bash", "/usr/src/start.sh" ]
```

### Environment variables

The following environment variables allow some degree of configuration:

| Environment variable | Description | Default | Options | 
| --- | --- | --- | --- |
| `AUDIO_LOG_LEVEL` | PulseAudio log level. | `WARN` | `ERROR`, `WARN`, `NOTICE`, `INFO`, `DEBUG`. |
| `AUDIO_OUTPUT` | Select the default audio output device. <br>Can also be changed at runtime by using the [companion library](#companion-library) | `AUTO` | For all device types: <br>- `AUTO`: Let PulseAudio decide. Priority is `USB > DAC > HEADPHONES > HDMI`<br>- `DAC`: Force default output to be an attached GPIO based DAC<br>- `<PULSE_SINK_NAME>`: If you know the sink name you can force set it too. Note that you can't use this to set custom sinks as default, in that case use `set-default-sink` on your custom pa script. <br><br> For Raspberry Pi devices: <br>- `RPI_AUTO`: BCM2835 automatic audio switching as described [here](https://web.archive.org/web/20200427023741/https://www.raspberrypi.org/documentation/configuration/audio-config.md). Deprecated for devices running Linux kernel 5.4 or newer. <br>- `RPI_HEADPHONES`: 3.5mm audio jack <br>- `RPI_HDMI0`: Main HDMI port <br>- `RPI_HDMI1`: Secondary HDMI port (only Raspberry Pi 4) <br><br> For Intel NUC: <br>- NUCs have automatic output detection and switching. If you plug both the HDMI and the 3.5mm audio jack it will use the latter.  |
| `AUDIO_VOLUME` | Initial volume level for the default audio output device. | `75%` | Any value between 0-100%. |

### Companion library

If you need to manipulate the block's behavior at runtime you can connect to the PulseAudio server, send commands and receive data or events from it. You should be able to use any existing library that implements the PA client protocol over TCP/UNIX sockets (some examples: [Python](https://pypi.org/project/pulsectl/), [Rust](https://docs.rs/libpulse-binding/2.16.0/libpulse_binding/), [JavaScript](https://github.com/stanford-oval/node-pulseaudio#readme)), or you could even [write your own](https://freedesktop.org/software/pulseaudio/doxygen/). Libraries that manipulate PA over dbus won't work because we don't run the pulse dbus daemon. 

On this note, we built a companion javascript library that exposes the most common use cases with an easy to use interface, these include changing the volume, listening to play/stop events, etc. 
You can install it with ```npm install balena-audio```; see https://github.com/balena-io-modules/balena-audio for usage and examples.

### Bluetooth

Bluetooth support for PulseAudio is enabled out of the box. Note that this only provides the backend that routes bluetooth packets over to PulseAudio, this does not include the Bluetooth agent that's required for initiating a connection and pairing devices. Check out our [Bluetooth block](https://github.com/balena-labs-projects/bluetooth) for an easy to use Bluetooth agent.

## Supported devices
The audio block has been tested to work on the following devices:

| Device Type  | Supported interface (driver) |
| ------------- | ------------- |
| Raspberry Pi (v1 / Zero / Zero W) | - Audio jack (`bcm2835`): ✔ <br>- HDMI (`bcm2835`): ✔<br>- I2S DAC (`snd-rpi-simple`): ✔ <br>- USB (`snd-usb-audio`): ✔ |
| Raspberry Pi 2 | - Audio jack (`bcm2835`): ✔ <br>- HDMI (`bcm2835`): ✔<br>- I2S DAC (`snd-rpi-simple`): ✔ <br>- USB (`snd-usb-audio`): ✔ |
| Raspberry Pi 3 | - Audio jack (`bcm2835`): ✔ <br>- HDMI (`bcm2835`): ✔<br>- I2S DAC (`snd-rpi-simple`): ✔ <br>- USB (`snd-usb-audio`): ✔ |
| Raspberry Pi 4 | - Audio jack (`bcm2835`): ✔ <br>- HDMI (`bcm2835`): ✔<br>- I2S DAC (`snd-rpi-simple`): ✔ <br>- USB (`snd-usb-audio`): ✔ |
| Intel NUC  | - Audio jack (`snd_hda_intel`): ✔<br>- HDMI (`snd_hda_intel`): ✔<br>- USB (`snd-usb-audio`): ✔ |
| Jetson Nano <sup>1</sup> | - HDMI (`tegrahda`): ✘ <br>- I2S DAC (`tegrasndt210ref`): ?<sup>2</sup> <br>- USB (`snd-usb-audio`): ✔ |
| BeagleBone Black | - USB (`snd-usb-audio`): ✔ |

1: Audio block crashes if no USB/DAC present. See: https://github.com/balena-labs-projects/audio/issues/35
2: Not tested. PR's welcome.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.5.6
## (2022-11-15)

* moving from balenablocks to balena-labs-projects [Flynn Joffray]

# v0.5.5
## (2022-03-29)

* meta: rename compose file [Tomás Migone]

# v0.5.4
## (2022-03-25)

* Update README.md [Tomás Migone]

# v0.5.3
## (2022-03-25)

* Update README.md [Tomás Migone]

# v0.5.2
## (2022-03-25)

* meta: use balenaHub container registry instead of docker [Tomás Migone]

# v0.5.1
## (2022-01-21)

* meta: force versionbot to run [Tomás Migone]
* defaults: fix default output and volume not being set properly [Tomás Migone]
* balena-audio: remove lib code, update readme [Tomás Migone]
* pa: disable dbus modules [Tomás Migone]
* pa: disable dbus modules [Tomás Migone]

# v0.5.0
## (2022-01-07)

* pulseaudio: bump to version 15 [Tomás Migone]

# v0.4.1
## (2021-12-23)

* bridge: replace install_packages with apk/apt [Tomás Migone]

# v0.4.0
## (2021-09-02)

* env: add variable to control init volume for default output [Tomás Migone]

# v0.3.0
## (2021-09-02)

* env: added variable to disable init logs [Tomás Migone]

# v0.2.4
## (2021-06-30)

* meta: remove CODEOWNERS file [Tomás Migone]

# v0.2.3
## (2021-05-31)

* Build blocks by arch type [Rahul Thakoor]

# v0.2.2
## (2021-05-27)

* lib: add connected check before issuing commands to PA [Tomás Migone]

# v0.2.1
## (2021-05-19)

* Rename primitive to block [Tomás Migone]

# v0.2.0
## (2021-05-19)

* base image: use alpine instead of debian [Tomás Migone]

# v0.1.13
## (2021-05-17)

* meta: update logo [Tomás Migone]
* rpi: use alpine base images for raspberry-pi images [Tomás Migone]

# v0.1.12
## (2021-03-05)

* pulseaudio: pin version to v13 [Tomás Migone]

# v0.1.11
## (2021-03-01)

* meta: force versionist to update balena.yml version [Tomás Migone]

# v0.1.10
## (2021-02-26)

* meta: sync balena.yml version [Tomás Migone]

# v0.1.9
## (2021-02-23)

* versionist: test versionist with balena.yml version specified [Tomás Migone]

# v0.1.8
## (2021-02-23)

* versionist: test new version in balena.yml contract [Tomás Migone]

# v0.1.7
## (2021-02-03)

* Update balena.yml [Tomás Migone]

# v0.1.6
## (2021-02-02)

* examples: update PULSE_SINK, don't specify one unless required [Tomás Migone]

# v0.1.5
## (2021-01-29)

* meta: Add files for hub [Tomás Migone]

# v0.1.4
## (2021-01-01)

* meta: add tinkerboard builds [Aaron Shaw]

# v0.1.3
## (2021-01-01)

* udev rules: add more driver names [Aaron Shaw]
* scripts: typo fix on comment [Aaron Shaw]

# v0.1.2
## (2020-12-01)

* alsa-bridge: Improve installation command, add success message [Tomás Migone]

# v0.1.1
## (2020-11-27)

* meta: build generic x86 image [Tomás Migone]

# v0.1.0
## (2020-11-13)

* meta: test versionbot [Tomás Migone]
* meta: add codeowners [Tomás Migone]
* meta: enable versionbot [Tomás Migone]
* raspberrypi: Fixes for devices running linux kernel 5.4 [Tomás Migone]
* usb: Fixed bug where sink names were incorrectly built for USB devices. [Tomás Migone]
* Add fincm3 images [Tomás Migone]
* images: add raspberrypi3-64 to build script [Tomás Migone]
* meta: move to balenablocks dockerhub org [Tomás Migone]
* docs: primitive --> block [Tomás Migone]
* meta: add config to publish package to GitHub Registry. [Tomás Migone]
* Fix bug with default sink setting. [Tomás Migone]
* fix bug where setting PA_SINK would break pa config filee [Tomás Migone]
* lib: update example to ts [Tomás Migone]
* audio: Don't rely on pulse default sink naming, can be inconsistent. Use USB > DAC > Built-in [Tomás Migone]
* Add support for jetson-nano and beaglebone-black [Tomás Migone]
* better error handling [Tomás Migone]
* bump dependencies [Tomás Migone]
* Better retry options for connection retries [Tomás Migone]
* Ensure /run/pulse exists on startup [Tomás Migone]
* Remove cookie option as the new default for audio block is to allow anonymous auth [Tomás Migone]
* lib: add log on client connect retry fail [Tomás Migone]
* Update dependencies [Tomás Migone]
* Allow anonymous authentication without cookie [Tomás Migone]
* lib: add retry to connect method [Tomás Migone]
* Rename primitive to block [Tomás Migone]
* Save default sink name to a temp file [Tomás Migone]
* Migrate all PA config to config files instead of cli flags [Tomás Migone]
* Update README.md [Tomás Migone]
* Add alpine alsa-bridge script, add rpi1 and 2 to build script [Tomás Migone]
* pa: logging default to notice [Tomás Migone]
* pa: Use udev rules for card naming [Tomás Migone]
* pa: output selector and flag changes [Tomás Migone]
* bluetooth: extract bluetooth primitive [Tomás Migone]
* images: add script to build and publish test images [Tomás Migone]
* Minor fixes [Tomás Migone]
* lib: Add primitive js library [Tomás Migone]
* pulse: Add option to pass in pulse cookie [Tomás Migone]
* README typos [Tomás Migone]
* Add ALSA bridge example and documentation [Tomás Migone]
* examples: Bump node version in nodejs example [Tomás Migone]
* meta: rearange repo folders [Tomás Migone]
* docs: typo [Tomás Migone]
* Multiple changes: - bluetooth: add BALENA_BLUETOOTH prefix to env vars - docs: add bluetooth section [Tomás Migone]
* docs: Add examples and info to readme file [Tomás Migone]
* pulse: add BALENA_AUDIO prefix to public facing env vars [Tomás Migone]
* readme: add env vars [Tomás Migone]
* multiple changes [Tomás Migone]
* pa: remove dockerfile tests [Tomás Migone]
* pulseaudio: add ALSA output selector [Tomás Migone]
* libs: add multiple pa libs to test [Tomás Migone]
* pulseaudio: use /run/pulse as default location for cookie file [Tomás Migone]
* pulseaudio: fix bug where primitive.pa conf file was not being loaded [Tomás Migone]
* dockerfiles: moved to use standard entrypoint [Tomás Migone]
* pulseaudio: refactor entry script [Tomás Migone]

entry.sh
#!/bin/bash
set -e

# Run balena base image entrypoint script
/usr/bin/entry.sh echo ""

# Helper functions
function pa_disable_module() {
  local MODULE="$1"
  sed -i "s/load-module $MODULE/#load-module $MODULE/" /etc/pulse/default.pa
}

function pa_set_log_level() {
  local PA_LOG_LEVEL="$1"
  declare -A options=(["ERROR"]=0 ["WARN"]=1 ["NOTICE"]=2 ["INFO"]=3 ["DEBUG"]=4)
  if [[ "${options[$PA_LOG_LEVEL]}" ]]; then
    LOWER_LOG_LEVEL=$(echo "$PA_LOG_LEVEL" | tr '[:upper:]' '[:lower:]')
    sed -i "s/log-level = notice/log-level = $LOWER_LOG_LEVEL/g" /etc/pulse/daemon.conf
  fi
}

function pa_set_cookie() {
  local PA_COOKIE="$1"
  if [[ ${#PA_COOKIE} == 512 && "$PA_COOKIE" =~ ^[0-9A-Fa-f]{1,}$ ]]; then
    echo "$PA_COOKIE" | xxd -r -p | tee /run/pulse/pulseaudio.cookie > /dev/null
  fi
}

function pa_read_cookie () {
  if [[ -f /run/pulse/pulseaudio.cookie ]]; then
    xxd -c 512 -p /run/pulse/pulseaudio.cookie
  fi
}

function pa_set_default_output () {
  local OUTPUT="$1"
  local PA_SINK=""

  # Note on RPi boards:
  # - RPI_ keys also act as indexes for setting the PCM Route for Raspberry Pi devices before 5.4 kernel, don't change the order or value of them.
  # - balenaOS v2.54.3+rev1 upgraded Linux kernel to 5.4 for Raspberry Pi 4 devices
  # - Docs for pre 5.4 --> https://web.archive.org/web/20200427023741/https://www.raspberrypi.org/documentation/configuration/audio-config.md
  declare -A options=(
    ["RPI_AUTO"]=0          # Deprecated on linux 5.4 
    ["RPI_HEADPHONES"]=1
    ["RPI_HDMI0"]=2
    ["RPI_HDMI1"]=3
    ["AUTO"]=4
    ["DAC"]=5
  )

  # Check /proc/asound for known cards
  BCM2835_CARDS=($(cat /proc/asound/cards | mawk -F '\[|\]:' '/bcm2835/ && NR%2==1 {gsub(/ /, "", $0); print $2}'))
  USB_CARDS=($(cat /proc/asound/cards | mawk -F '\[|\]:' '/usb/ && NR%2==1 {gsub(/ /, "", $0); print $2}'))
  DAC_CARD=$(cat /proc/asound/cards | mawk -F '\[|\]:' '/dac|DAC|Dac/ && NR%2==1 {gsub(/ /, "", $0); print $2}')
  HDA_CARD=$(cat /proc/asound/cards | mawk -F '\[|\]:' '/hda-intel/ && NR%2==1 {gsub(/ /, "", $0); print $2}')

  case "${options[$OUTPUT]}" in

    # RPi familiy
    ${options["RPI_AUTO"]} | ${options["RPI_HEADPHONES"]} | ${options["RPI_HDMI0"]} | ${options["RPI_HDMI1"]})
      if [[ -n "$BCM2835_CARDS" ]]; then
        if [[ "${BCM2835_CARDS[@]}" =~ "bcm2835-alsa" ]]; then
          # Devices running linux kernel < 5.4
          amixer --card bcm2835-alsa --quiet cset numid=3 "${options[$OUTPUT]}"
          PA_SINK="alsa_output.bcm2835-alsa.stereo-fallback"
        else
          # Devices running linux kernel >= 5.4
          if [[ "${options[$OUTPUT]}" == "${options["RPI_HEADPHONES"]}" ]]; then
            PA_SINK="alsa_output.bcm2835-jack.stereo-fallback"
          elif [[ "${options[$OUTPUT]}" == "${options["RPI_HDMI0"]}" ]]; then
            PA_SINK="alsa_output.bcm2835-hdmi0.stereo-fallback"
          elif [[ "${options[$OUTPUT]}" == "${options["RPI_HDMI1"]}" ]]; then
            PA_SINK="alsa_output.bcm2835-hdmi1.stereo-fallback"
          else
            echo "WARNING: Option not supported for this kernel version. Using defaults..."
          fi
        fi
      else
        echo "WARNING: BCM2835 audio card not found, are you sure you are running on a Raspberry Pi?"
      fi
      ;;

    # DACs
    ${options["DAC"]})
      if [[ -n "$DAC_CARD" ]]; then
        PA_SINK="alsa_output.dac.stereo-fallback"
      else
        echo "WARNING: No DAC found. Falling back to PulseAudio defaults."
      fi
      ;;

    # AUTO - USB > DAC > BUILT-IN
    ${options["AUTO"]})
      declare -a sound_cards=("${USB_CARDS[@]}" "$DAC_CARD" "${BCM2835_CARDS[@]}")
      for sound_card in "${sound_cards[@]}"
      do
        if [[ -n "$sound_card" ]]; then
          if [[ -n "$USB_CARDS" ]]; then
            PA_SINK="alsa_output.${USB_CARDS[0]}.analog-stereo"
          elif [[ -n "$DAC_CARD" ]]; then
            PA_SINK="alsa_output.dac.stereo-fallback"
          elif [[ -n "$BCM2835_CARDS" ]]; then
            if [[ "${BCM2835_CARDS[@]}" =~ "bcm2835-alsa" ]]; then
              # Devices running linux kernel < 5.4
              PA_SINK="alsa_output.bcm2835-alsa.stereo-fallback"
            else
              # Devices running linux kernel >= 5.4
              PA_SINK="alsa_output.bcm2835-jack.stereo-fallback"
            fi
          fi
          break
        fi
      done
      ;;

    # If there was no match, we asume the provided value is the name of a PulseAudio sink.
    *)
      PA_SINK="$OUTPUT"
      ;;
  esac

  # Set the sink name as PA default and save it in a temp file
  if [[ -n "$PA_SINK" ]]; then
    echo "$PA_SINK" > /run/pulse/pulseaudio.sink
    echo -e "\nset-default-sink $PA_SINK" >> /etc/pulse/default.pa.d/00-audioblock.pa
  fi
}

# Platform specific commands to initialize audio hardware
function init_audio_hardware () {
  # Allow hardware to be initialized, PulseAudio only creates cards/sinks/sources on startup
  sleep 10

  # Intel NUC
  HDA_CARD=$(cat /proc/asound/cards | mawk -F '\[|\]:' '/hda-intel/ && NR%2==1 {gsub(/ /, "", $0); print $2}')
  if [[ -n "$HDA_CARD" ]]; then
    amixer --card hda-intel --quiet cset numid=2 on,on  # Master Playback Switch --> turn on hardware
    amixer --card hda-intel --quiet cset numid=1 87,87  # Master Playback Volume --> max volume
    PA_SINK="alsa_output.hda-intel.analog-stereo"
  fi
}

function print_audio_cards () {
  cat /proc/asound/cards | mawk -F '\[|\]:' 'NR%2==1 {gsub(/ /, "", $0); print $1,$2,$3}'
}

function sanitize_volume () {
  local VOLUME="${1//%}"
  if [[ "$VOLUME" -ge 0 && "$VOLUME" -le 100 ]]; then
    echo "$VOLUME"
  fi
}

function pa_set_default_volume () {
  local VOLUME_PERCENTAGE=$(sanitize_volume "$1")
  local VOLUME_ABSOLUTE=$(( VOLUME_PERCENTAGE * 65536 / 100 ))
  if [[ -n "$VOLUME_ABSOLUTE" ]]; then
    echo -e "\nset-sink-volume @DEFAULT_SINK@ $VOLUME_ABSOLUTE" >> /etc/pulse/default.pa.d/00-audioblock.pa
  fi
}

# PulseAudio block environment variables and defaults
INIT_LOG="${AUDIO_INIT_LOG:-true}"
LOG_LEVEL="${AUDIO_LOG_LEVEL:-NOTICE}"
COOKIE="${AUDIO_PULSE_COOKIE}"
DEFAULT_OUTPUT="${AUDIO_OUTPUT:-AUTO}"
DEFAULT_VOLUME="${AUDIO_VOLUME:-75}"

if [[ "$INIT_LOG" != "false" ]]; then
  echo "--- Audio ---"
  echo "Starting audio service with settings:"
  echo "- $(pulseaudio --version)"
  echo "- Pulse log level: $LOG_LEVEL"
  [[ -n ${COOKIE} ]] && echo "- Pulse cookie: $COOKIE"
  echo "- Default output: $DEFAULT_OUTPUT"
  echo "- Default volume: $DEFAULT_VOLUME%"
  echo -e "\nDetected audio cards:"
  print_audio_cards
  echo -e "\n"
fi

# Create dir for temp/share files
mkdir -p /run/pulse

# Configure audio hardware
# ALSA CONFIG
init_audio_hardware
pa_set_default_output "$DEFAULT_OUTPUT"
pa_set_default_volume "$DEFAULT_VOLUME"

# PULSE AUDIO CONFIG
# Disable PulseAudio modules that we don't support
pa_disable_module module-console-kit
pa_disable_module module-dbus-protocol
pa_disable_module module-jackdbus-detect

# Disable PulseAudio modules that need special configuration
# These will be loaded and configured by /etc/pulse/default.pa.d/00-audioblock.pa
pa_disable_module module-bluetooth-discover
pa_disable_module module-bluetooth-policy
pa_disable_module module-native-protocol-unix

pa_set_log_level "$LOG_LEVEL"

# Set PulseAudio cookie
if [[ -n "$COOKIE" ]]; then
  pa_set_cookie "$COOKIE"
fi

# If command starts with an option, prepend PulseAudio to it
if [[ "${1#-}" != "$1" ]]; then
  set -- pulseaudio "$@"
fi

exec "$@"